Remote attestation PTA

jforissier opened the pull request:
Differences with [v4](https://github.com/OP-TEE/optee_os/pull/5010):
- The hash of TA memory pages from [v2](https://github.com/OP-TEE/optee_os/pull/4972) is back, in addition to obtaining the digest from the TA secure header (struct shdr::hash) is still possible. This way more use cases can be supported.
- Attestation of TEE core memory is added, but see the limitation with ASLR in the commit description (might or might not be an big issue depending on the use case).

~~Feel free to formally review the commits that are not tagged "[RFC]", although they are needed only for the attestation PTA, they also make sense on their own (except perhaps the first one "scripts: add dump_ta_header.py" and last one "kernel/linker.h: export __text_end").~~

Update: "RFC v5" tag removed.

jenswi-linaro commented on core/tee/tee_fs_rpc.c:
```diff
@@ -23,6 +24,27 @@ struct tee_fs_dir {
 
+/* "/dirf.db" or "/<file number>" */
+static TEE_Result
+tee_svc_storage_create_filename_dfh(void *buf, size_t blen,
```
Would it make sense to rename this function to `create_filename()` also?

jforissier replied:
Absolutely.

jenswi-linaro commented on core/tee/tee_svc_storage.c:
```diff
@@ -104,29 +103,6 @@ TEE_Result tee_svc_storage_create_filename(void *buf, size_t blen,
-#endif
-
 /* "/TA_uuid" */
 TEE_Result tee_svc_storage_create_dirname(void *buf, size_t blen,
```
How about moving this function and `tee_svc_storage_create_filename()` to `tee_rpmb_fs.c` following the same logic?

jforissier replied:
Yes, will do.

jenswi-linaro reviewed (commented):
Comments for "core: move tee_svc_storage_create_filename_dfh() to core/tee/tee_fs_rpc.c"

jenswi-linaro commented on core/tee/sub.mk:
```diff
@@ -24,25 +24,29 @@ srcs-$(CFG_CRYPTO_CONCAT_KDF) += tee_cryp_concat_kdf.c
 
+ifeq ($(_CFG_WITH_SECURE_STORAGE),y)
+srcs-y += tee_fs_key_manager.c
+srcs-$(CFG_RPMB_FS) += tee_rpmb_fs.c
```
Given that `_CFG_WITH_SECURE_STORAGE` will be `y` if `CFG_RPMB_FS` is `y` it seems unnecessary to have this inside the if.

jforissier replied:
Correct, I will simplify.

jenswi-linaro reviewed (commented):
Comment for "core: move FS files under _CFG_WITH_SECURE_STORAGE"

jforissier commented:
Comments addressed.

jenswi-linaro commented:
For the following commits with their fixups please apply:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jforissier commented:
Fixup commits squashed and review tag applied. I have changed the subject of "core: move tee_svc_storage_create_filename() ..." and improved the description too, it is now "core: move functions from tee_svc_storage.c to tee_rpmb_fs.c".

jenswi-linaro commented:
I think it would make sense to merge these reviewed commits separately. I should also make it easier to focus on what this PR is about.

jforissier commented:
> I think it would make sense to merge these reviewed commits separately. I should also make it easier to focus on what this PR is about.

Sure. I have merged them just now and rebased the branch. Thanks for reviewing.

jenswi-linaro commented on core/arch/arm/kernel/kern.ld.S:
```diff
@@ -93,6 +93,9 @@ SECTIONS
 		KEEP(*(.text._start))
 		__identity_map_init_start = .;
+		__text_data_start = .;
+		*(.text.data)
```
If there's ever a function, static or global, named data it will go into this too.
How about using `.identify_map.data` instead? That way we'd have full control of what's going in here.

jforissier replied:
Fair enough, and it also makes it obvious that the data need to be reachable using the identity mapping which is good.
Can I still keep `__text_data_start`/`__text_data_end`? To the "outside world" (tools etc.) it's essentially data embedded in `.text`.

jenswi-linaro replied:
Sure

jenswi-linaro commented:
Please squash the fixups.

jforissier commented:
> Please squash the fixups.

Will do. Some more work is needed to deal with the ASLR case, because there is still relocatable data in both the `.text` and `.rodata` sections. I'll refrain from pushing to this branch until I have something that works and is in a good shape to be reviewed. Thanks for the comments so far.

jforissier commented:
Update. The attestation commit is mostly unchanged but I have added a number of commits to suppress text relocations and a few more to avoid storing relocatable data into `.rodata`. @jenswi-linaro despite your advice to use the stack, I couldn't figure out how to deal with the transition from reduced to full mapping though :-/ please see commit "FIXME: core: arm: disable CFG_UNMAP_CORE_AT_EL0 by default". I wouldn't mind some more help on that one...

jenswi-linaro commented:
Perhaps something like (untested) this would work. This stuff is a bit hairy so let me know if you'd rather I did this.

``` diff
diff --git a/core/arch/arm/include/kernel/thread.h b/core/arch/arm/include/kernel/thread.h
index d2b0d006c8de..ce8069fbec25 100644
--- a/core/arch/arm/include/kernel/thread.h
+++ b/core/arch/arm/include/kernel/thread.h
@@ -43,6 +43,7 @@ struct thread_core_local {
 	uint64_t x[4];
 #endif
 	vaddr_t tmp_stack_va_end;
+	long kcode_offset;
 	short int curr_thread;
 	uint32_t flags;
 	vaddr_t abt_stack_va_end;
diff --git a/core/arch/arm/kernel/asm-defines.c b/core/arch/arm/kernel/asm-defines.c
index 76c87de881a0..c1436baa937c 100644
--- a/core/arch/arm/kernel/asm-defines.c
+++ b/core/arch/arm/kernel/asm-defines.c
@@ -94,6 +94,8 @@ DEFINES
 	/* struct thread_core_local */
 	DEFINE(THREAD_CORE_LOCAL_X0, offsetof(struct thread_core_local, x[0]));
 	DEFINE(THREAD_CORE_LOCAL_X2, offsetof(struct thread_core_local, x[2]));
+	DEFINE(THREAD_CORE_LOCAL_KCODE_OFFSET,
+	       offsetof(struct thread_core_local, kcode_offset));
 #endif /*ARM64*/
 
 	/* struct thread_ctx */
diff --git a/core/arch/arm/kernel/thread.c b/core/arch/arm/kernel/thread.c
index 05dd8697e856..e89d85092c48 100644
--- a/core/arch/arm/kernel/thread.c
+++ b/core/arch/arm/kernel/thread.c
@@ -1045,6 +1045,15 @@ static void init_thread_stacks(void)
 }
 #endif /*CFG_WITH_PAGER*/
 
+static void set_core_local_kcode_offset(struct thread_core_local *cls,
+					long offset)
+{
+	size_t n = 0;
+
+	for (n = 0; n < CFG_TEE_CORE_NB_CORE; n++)
+		cls[n].kcode_offset = offset;
+}
+
 static void init_user_kcode(void)
 {
 #ifdef CFG_CORE_UNMAP_CORE_AT_EL0
@@ -1058,7 +1067,11 @@ static void init_user_kcode(void)
 	core_mmu_get_user_va_range(&v, NULL);
 	thread_user_kcode_offset = thread_user_kcode_va - v;
 
+	set_core_local_kcode_offset(thread_core_local,
+				    thread_user_kcode_offset);
 #if defined(CFG_CORE_WORKAROUND_SPECTRE_BP_SEC) && defined(ARM64)
+	set_core_local_kcode_offset((void *)thread_user_kdata_page,
+				    thread_user_kcode_offset);
 	/*
 	 * When transitioning to EL0 subtract SP with this much to point to
 	 * this special kdata page instead. SP is restored by add this much
diff --git a/core/arch/arm/kernel/thread_a64.S b/core/arch/arm/kernel/thread_a64.S
index 5d70d9ed24fe..9d13f81ae327 100644
--- a/core/arch/arm/kernel/thread_a64.S
+++ b/core/arch/arm/kernel/thread_a64.S
@@ -188,7 +188,13 @@ END_FUNC thread_unwind_user_mode
 		isb
 
 		/* Jump into the full mapping and continue execution */
+#if 0
 		ldr	x0, =1f
+#else
+		adr	x0, 1f
+		ldr	x1, [sp, #THREAD_CORE_LOCAL_KCODE_OFFSET]
+		add	x0, x0, x1
+#endif
 		br	x0
 	1:
 BTI(		bti	j)
@@ -484,7 +490,11 @@ eret_to_el0:
 #endif
 
 	/* Jump into the reduced mapping and continue execution */
+#if 0
 	ldr	x1, =1f
+#else
+	adr	x1, 1f
+#endif
 	sub	x1, x1, x0
 	br	x1
 1:
```

jforissier commented:
@jenswi-linaro awesome, your suggestion works fine and eliminates 6 relocations in `thread_a64.S`. There is one left however :-/
In `icache_inv_user_range()`, there is a `ldr x0, =1f`. I suspect the same trick can be applied, but how to find the location if the offset? Not based on `sp` at this point, right?

jenswi-linaro commented:
This (untested) code might work:
``` diff
diff --git a/core/arch/arm/kernel/thread_a64.S b/core/arch/arm/kernel/thread_a64.S
index 5d70d9ed24fe..1356fd53304f 100644
--- a/core/arch/arm/kernel/thread_a64.S
+++ b/core/arch/arm/kernel/thread_a64.S
@@ -535,16 +535,16 @@ icache_inv_user_range:
 
 #ifdef CFG_CORE_UNMAP_CORE_AT_EL0
 	/* Point to the vector into the reduced mapping */
-	adr_l	x2, thread_user_kcode_offset
-	ldr	x2, [x2]
+	adr_l	x7, thread_user_kcode_offset
+	ldr	x7, [x7]	/* this register must be preserved */
 	mrs	x4, vbar_el1	/* this register must be preserved */
-	sub	x3, x4, x2
+	sub	x3, x4, x7
 	msr	vbar_el1, x3
 	isb
 
 	/* Jump into the reduced mapping and continue execution */
-	ldr	x3, =1f
-	sub	x3, x3, x2
+	adr	x3, 1f
+	sub	x3, x3, x7
 	br	x3
 1:
 BTI(	bti	j)
@@ -588,7 +588,8 @@ BTI(	bti	j)
 	isb
 
 	/* Jump into the full mapping and continue execution */
-	ldr	x0, =1f
+	adr	x0, 1f
+	add	x0, x0, x7
 	br	x0
 1:
 BTI(	bti	j)
```

jforissier commented:
> This (untested) code might work:

Guess what? It does! :+1: I will introduce theses changes at the right place in the series. Thanks again!

jenswi-linaro commented on core/arch/arm/kernel/thread_a64.S:
```diff
@@ -535,16 +535,16 @@ icache_inv_user_range:
 	/* Jump into the reduced mapping and continue execution */
-	ldr	x3, =1f
-	sub	x3, x3, x2
+	adr_l	x3, 1f
```
A plain `adr` should be enough, same below. This works because it's not very far away.

jforissier replied:
Correct, fixed.

jenswi-linaro commented on core/arch/arm/plat-rcar/romapi_call.S:
```diff
@@ -31,9 +32,11 @@ FUNC __plat_romapi_wrapper , : , .identity_map
+	 * which is not identity mapped.
 	 */
-	ldr	x9, =virt_to_phys
+	adr     x9, virt_to_phys
```
This should probably be an `adr_l` instead since we don't know how far away it might be, same below.

jforissier replied:
Yes, will fix.

jenswi-linaro reviewed (commented):
Comment for "plat-rcar: suppress text relocations in romapi_call.S"

jenswi-linaro commented on core/arch/arm/kernel/kern.ld.S:
```diff
@@ -234,6 +232,7 @@ SECTIONS
 		__data_start = .;
 		*(.data .data.* .gnu.linkonce.d.*)
 		. = ALIGN(8);
+		KEEP(*(SORT(.scattered_array*)));
```
This will prevent the scattered arrays from being paged any longer. Another drawback is that they will become less protected at runtime since they are now mapped read-write.

jenswi-linaro reviewed (commented):
Comments for "core: kern.ld.S: move .scattered_array* into .data"

jenswi-linaro commented on core/arch/arm/kernel/kern.ld.S:
```diff
@@ -233,6 +228,12 @@ SECTIONS
 		. = ALIGN(8);
 		KEEP(*(SORT(.scattered_array*)));
+#ifdef CFG_DT
+		. = ALIGN(8);
```
Same problem here as with scattered arrays. While looking at this I wonder why the dtdrv stuff isn't implemented using scattered arrays.

jenswi-linaro reviewed (commented):
Comment for "core: kern.ld.S: rename .rodata.dtdrv to .data.dtdrv, move to .data"

jenswi-linaro commented:
With this we actually removes some protection by making data that where supposed to remain constant after relocation read-write mapped instead. Perhaps we instead should collect all this data in a special section that can be mapped read-only and still be paged.

jforissier commented:
> With this we actually removes some protection by making data that where supposed to remain constant after relocation read-write mapped instead. Perhaps we instead should collect all this data in a special section that can be mapped read-only and still be paged.

That's a valid concern indeed. I will think about it.

jforissier commented:
Rebased onto master which gets rid of the dtdrv stuff. New commits added to address comments are 8e403217 "[Review] core: kern.ld.S: move .scattered_array* into .data" and later.

jenswi-linaro commented on core/arch/arm/kernel/kern.ld.S:
```diff
@@ -134,11 +134,21 @@ SECTIONS
 		__rodata_end = .;
 	}
 
+#if !defined(CFG_WITH_PAGER) && defined(CFG_CORE_ASLR)
```
Why can't this always be used even when ASLR isn't enabled?

jforissier replied:
It can, but doesn't have to. I thought having a section called `.data.rel.ro` could be slightly misleading when ASLR is disabled (and therefore the executable is not relocatable).
I can simplify if you prefer.

jenswi-linaro replied:
Makes sense, let's keep it as it is now.

b49020 commented:
@jforissier Thanks again for clarifying queries at OP-TEE LOC meeting. I think from overall design aspect, this feature looks in better shape now. So we should be good to remove RFC tag at this point. If anyone else has different views, please feel free to comment.

Also, can you squash all the review and update commits? As I would like to spend some time to review and test actual code afresh.

BTW, if anybody missed OP-TEE LOC meeting discussion regarding this feature, [here](https://docs.google.com/presentation/d/10vrFfxCt7Th8OL05jq-_2-uNVaJXQaH3omP97AsCVy4/edit#slide=id.p) is the reference link to the slides.

jforissier commented:
@b49020 yes I will squash the updates and rebase onto master. I noticed I forgot to hash `.text_init` and `.rodata_init` (present when pager is enabled), I will fix that too.

b49020 commented:
While looking at this patch-set, following commit caught mine attention.

`Commit: "core: do not tag ops structures with __rodata_unpaged"`

It looks like after this commit with `CFG_CORE_ASLR=y`, we are essentially allowing constant operations structs writable. IOW, no MMU RO protection which was there earlier with `__rodata_unpaged`. This will make it easier for any vulnerability found in TEE to modify function pointers and redirect code flow. Although having measurement for these ops structs included would be the best solution but this is where ASLR makes it difficult (two security features limiting each other's capability rather than attacker's :)).

However, we shouldn't remove MMU RO protection for these ops struct and atleast can have a special section for them to be excluded from measurement. Any better ideas are much appreciated.

b49020 reviewed (commented):
This review doesn't include `core/pta/attestation.c` as that's something I will look into next.

jforissier commented:
> While looking at this patch-set, following commit caught mine attention.
> 
> `Commit: "core: do not tag ops structures with __rodata_unpaged"`
> 
> It looks like after this commit with `CFG_CORE_ASLR=y`, we are essentially allowing constant operations structs writable. IOW, no MMU RO protection which was there earlier with `__rodata_unpaged`. This will make it easier for any vulnerability found in TEE to modify function pointers and redirect code flow. Although having measurement for these ops structs included would be the best solution but this is where ASLR makes it difficult (two security features limiting each other's capability rather than attacker's :)).
> 
> However, we shouldn't remove MMU RO protection for these ops struct and atleast can have a special section for them to be excluded from measurement. Any better ideas are much appreciated.

You're right. In fact these structures are quite similar to the scattered arrays, so we can probably store them in `.data.rel.ro*` when ASLR is enabled. I will propose something.

jforissier commented on core/arch/arm/kernel/kern.ld.S:
```diff
@@ -323,13 +331,18 @@ SECTIONS
+	}
+#endif
+	. = ALIGN(8);
+	__rodata_init_end = .;
```
Good catch ;-) This is actually fixed by later commit "core: pager: export __{text,rodata}_{init,pageable}_{start,end}". At this point it doesn't matter since the attestation PTA doesn't exist yet.
Still I suppose it would make more sense to introduce `__ro_and_relro_data_init_end` at the same time as the new relro section. I will fix that.

jforissier commented on mk/config.mk:
```diff
@@ -574,6 +574,17 @@ $(eval $(call cfg-depends-all,CFG_SYSTEM_PTA,CFG_WITH_USER_TA))
 
+# The attestation pseudo TA provides an interface to request a measurement of
+# a given TA binary or shared library given its UUID.
+CFG_ATTESTATION_PTA ?= $(_CFG_WITH_SECURE_STORAGE)
```
Maybe. I would like it to be tested by CI, but I can `export CFG_ATTESTATION_PTA=y` in `.azure-pipelines.yml` instead.

jforissier commented:
@b49020 I have fixed your comments in place rather than add fixup commits because it would have been hard to understand. Also rebased on master.
Changes since last review:
- Introduce __ro_and_relro_data_init_end in "core: kern.ld.S: move .scattered_array* into .data.rel.ro" (note: subject previously said '.data', it was a mistake)
- Discard "core: do not tag ops structures with __rodata_unpaged"
- Instead, add "core: compiler.h: introduce __relrodata_unpaged(x)" and "core: tag ops structures with __relrodata_unpaged". Question: is it OK to still use __rodata_dummy in `link_dummies_paged.c`?
- Fix a small issue in print_tee_hash.py: do not fail when a section is not found, it does happen when pager is disabled(commit: "scripts: add print_tee_hash.py").

Tested with and without pager, with and without ASLR.

b49020 commented:
@jforissier Thanks for the update.

For commits:
```
core: suppress text relocation in thread.c
core: arm64: suppress text relocations in thread_optee_smc_a64.S
core: arm64: suppress some text relocations in thread_a64.S
core: arm64: suppress text relocations in icache_inv_user_range()
core: arm64: suppress text relocations caused by restore_mapping macro
plat-rcar: suppress text relocations in romapi_call.S
kernel/linker.h: export __text_end
core: add delimited area in .text to store data
core: arm64: move boot_mmu_config and cached_mem_end to .identity_map.data
core: kern.ld.S: move .scattered_array* into .data.rel.ro
core: compiler.h: introduce __relrodata_unpaged(x)
core: tag ops structures with __relrodata_unpaged
core: arm64: aslr: pass '-z text' linker option
core: pager: export __{text,rodata}_{init,pageable}_{start,end}
```
feel free to add:
`Acked-by: Sumit Garg <sumit.garg@linaro.org>`

> Question: is it OK to still use __rodata_dummy in link_dummies_paged.c?

I don't see a point to have declarations separately in this file. So shouldn't those be dropped entirely?

b49020 commented on lib/libutee/include/pta_attestation.h:
```diff
@@ -0,0 +1,99 @@
+		{ 0x9b, 0x67, 0x2b, 0xcd, 0x62, 0x2b, 0xc0, 0xb5 } }
+
+/*
+ * Get the RSA public key use for signing the values returned by other commands.
```
s/signing/verification/

jforissier replied:
Right

b49020 commented on lib/libutee/include/pta_attestation.h:
```diff
@@ -0,0 +1,99 @@
+ * TEE_ERROR_SHORT_BUFFER - One or both buffers are too small, required size
+ *                          is provided in memref[i].size
+ */
+#define PTA_ATTESTATION_GET_PUBKEY 0x0
```
Although currently we are using RSA as the signing scheme but it should be possible in future to extend this pTA to use ECDSA scheme instead in a backwards compatible way. So I would suggest you to add an output value parameter to say that is an RSA public key.

jforissier replied:
I agree that other encryption schemes may be useful in the future, but I'd rather avoid introducing too much complexity now. I think what you are asking can easily be done with minimal code changes by adding a new command.

b49020 replied:
Yeah adding a new info command would be good alternative. I think it would be worth to add at this point. As it would avoid to hard code signature type and algorithm used in both pseudo TA and remote client when that info can be fetched at runtime.

jforissier replied:
How about `PTA_ATTESTATION_GET_PUBKEY` would return `TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA256` as `[out] value[2].a`?

b49020 replied:
That's fine with me as well.

b49020 commented on lib/libutee/include/pta_attestation.h:
```diff
@@ -0,0 +1,99 @@
+ * TEE_ERROR_BAD_PARAMETERS - Incorrect input param
+ * TEE_ERROR_SHORT_BUFFER - Output buffer size less than required
+ */
+#define PTA_ATTESTATION_GET_TA_SHDR_DIGEST 0x1
```
Here as well, we should have an output value parameter to say that its an RSA signature.

b49020 commented on lib/libutee/include/pta_attestation.h:
```diff
@@ -0,0 +1,99 @@
+ * TEE_ERROR_BAD_PARAMETERS - Incorrect input param
+ * TEE_ERROR_SHORT_BUFFER - Output buffer size less than required
+ */
+#define PTA_ATTESTATION_HASH_TA_MEMORY 0x2
```
Here as well, we should have an output value parameter to say that its an RSA signature.

b49020 commented on lib/libutee/include/pta_attestation.h:
```diff
@@ -0,0 +1,99 @@
+ * TEE_ERROR_BAD_PARAMETERS - Incorrect input param
+ * TEE_ERROR_SHORT_BUFFER - Output buffer size less than required
+ */
+#define PTA_ATTESTATION_HASH_TEE_MEMORY 0x3
```
Here as well, we should have an output value parameter to say that its an RSA signature.

b49020 commented on mk/config.mk:
```diff
@@ -574,6 +574,17 @@ $(eval $(call cfg-depends-all,CFG_SYSTEM_PTA,CFG_WITH_USER_TA))
+$(eval $(call cfg-depends-all,CFG_ATTESTATION_PTA,_CFG_WITH_SECURE_STORAGE))
+
+# RSA key size (in bits) for the attestation PTA. Must be at least 2048 which
+# is the size of the data to be signed (an SHA-256 hash).
```
> Must be at least 2048 which is the size of the data to be signed (an SHA-256 hash).

Isn't size of SHA-256 hash 256 bits? Although I agree that choosing 2048 as RSA key size is better default.

jforissier replied:
Actually what I wrote is complete BS :open_mouth: :smile: I will update the text and explain the key size constraint more precisely.

b49020 commented on lib/libutee/include/pta_attestation.h:
```diff
@@ -0,0 +1,99 @@
+ *
+ * Return codes:
+ * TEE_SUCCESS
+ * TEE_ERROR_ACCESS_DENIED - Caller is not a user space TA
```
Is this error code relevant here?

jforissier replied:
Good catch, I'll remove.

b49020 reviewed (commented):
Comments for pTA interface: `core: pta: add remote attestation PTA`

jforissier commented on mk/config.mk:
```diff
@@ -574,6 +574,17 @@ $(eval $(call cfg-depends-all,CFG_SYSTEM_PTA,CFG_WITH_USER_TA))
 CFG_DEVICE_ENUM_PTA ?= y
 
+# The attestation pseudo TA provides an interface to request a measurement of
+# a given TA binary or shared library given its UUID.
```
Will fix

jforissier commented:
Update

b49020 reviewed (commented):
pTA implementation comments.

jforissier commented on core/pta/attestation.c:
```diff
@@ -0,0 +1,761 @@
+{
+	assert(!key);
+
+	key = calloc(sizeof(*key), 1);
```
You're right, I will fix all occurrences.

jforissier commented on core/pta/attestation.c:
```diff
@@ -0,0 +1,761 @@
+	 * Serialized key pair is 3 bignums (e, p and n) plus their sizes
+	 * encoded as uint16_t. e is 65537 so it needs only 3 bytes.
+	 */
+	size_t size = 3 + 2 * MAX_KEY_SIZE / 8 + 3 * sizeof(uint16_t);
```
- I used `MAX_KEY_SIZE` rather than `CFG_ATTESTATION_PTA_KEY_SIZE` because I thought maybe we should be able to read a larger key that might have been generated and saved by a prior version of the TEE.
- I will move the formula to a macro (or even better a static function)
- I will also allocate the buffer within `init_key()` as suggested.

b49020 replied:
Makes sense to me.

jforissier commented on core/pta/attestation.c:
```diff
@@ -0,0 +1,761 @@
+
+	fops = tee_svc_storage_file_ops(storage_id);
+	if (!fops)
+		return TEE_ERROR_ITEM_NOT_FOUND;
```
Should not make a difference in practice, but "not implemented" does sound better.

jforissier commented on core/pta/attestation.c:
```diff
@@ -0,0 +1,761 @@
+	res = crypto_hash_init(ctx);
+	if (res)
+		goto out;
+	if (nonce) {
```
No, but it would make the documentation more complex :wink "SIG(SHA256(Nonce | digest)) except when no Nonce is provided in which case it is SIG(digest)". Perhaps I should simply reject `nonce == NULL`? It doesn't seem safe to run the algorithm without a nonce.

b49020 replied:
I agree, we should reject to perform a signature without nonce.

jforissier commented on core/pta/attestation.c:
```diff
@@ -0,0 +1,761 @@
+		return TEE_ERROR_SHORT_BUFFER;
+	out_sz = min_out_sz;
+
+	/* Check that we're called from a user TA */
```
Would be better yes.

b49020 commented:
With comments (https://github.com/OP-TEE/optee_os/pull/5025#discussion_r797596335, https://github.com/OP-TEE/optee_os/pull/5025#discussion_r798518264) addressed, feel free to add:

`Reviewed-by: Sumit Garg <sumit.garg@linaro.org>`

for commits:

```
scripts: add print_tee_hash.py
core: pta: add remote attestation PTA
```

jforissier commented:
@b49020 thanks for the review!

b49020 commented:
@jforissier you are welcome but still I don't this https://github.com/OP-TEE/optee_os/pull/5025#discussion_r797596335 addressed.

jforissier commented:
@b49020 sorry I forgot to add the change :-/ now done. I have also once again revised the explanation for the minimal key size, because I had referred to the wrong section in RFC8017...

jforissier commented:
Rebased to resolve conflicts.

jforissier commented:
If there is no more comment I will merge this. @jenswi-linaro?

jenswi-linaro commented on core/arch/arm/kernel/entry_a64.S:
```diff
@@ -29,8 +30,7 @@
 		mul	x1, x0, x1
-		adr_l	x0, stack_tmp_export
-		ldr	x0, [x0]
+		bl	thread_get_stack_tmp_export
```
SP is unassigned at this point. How can you know that `thread_get_stack_tmp_export()` doesn't use the stack?

jforissier replied:
Good point. It happens to be the case (I checked the disassembly), but there is obviously no guarantee so I need to find another way. I think I'll rewrite the function in assembly unless you have a better idea.

jenswi-linaro replied:
In that case we could just add the needed instructions here.

jforissier commented:
Update:
- Commit "core: suppress text relocations in thread.c" reverted and replaced with "core: suppress text relocation on stack_tmp_export"
- Rebased onto master

jenswi-linaro commented on core/arch/arm/kernel/entry_a32.S:
```diff
@@ -823,9 +829,9 @@ LOCAL_FUNC enable_mmu , : , .identity_map
-	.word	stack_tmp_export - stack_tmp_export_rel
-END_DATA stack_tmp_export_rel
+LOCAL_DATA stack_tmp_rel , :
+	.word	stack_tmp - stack_tmp_rel
```
I'm not sure but it seems that making this `stack_tmp - stack_tmp_rel - STACK_TMP_GUARD` should allow us to skip the entire `#if (STACK_TMP_GUARD != 0)` above.

jforissier replied:
It does work. Updated.

jenswi-linaro replied:
Great, please do the same in the _a64.S.

jforissier replied:
Done.

jenswi-linaro commented:
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

jforissier commented:
@b49020 @jenswi-linaro thanks for your time, I am merging this now.
